File: java-gradle/aibest/src/main/java/net/pijcke/nim/aibest/BestAI.java
package net.pijcke.nim.aibest;

import net.pijcke.nim.airandom.RandomAI;
import net.pijcke.nim.rules.Nim;
import net.pijcke.nim.rules.Player;

import java.util.Random;

public class BestAI implements Player {

    private final RandomAI randomAi;
    private static final int MAX_PLUS_ONE = Nim.MAX_STICKS_PER_TURN + 1;
    private static final int MAX = Nim.MAX_STICKS_PER_TURN;

    public BestAI() {
        randomAi = new RandomAI();
    }

    @Override
    public String name() {
        return "Best AI";
    }

    @Override
    public int play(int sticks) {
        if (sticks == 1) {
            return 1;
        }

        if (sticks % (MAX_PLUS_ONE) == 1) {
            return randomAi.play(sticks);
        }

        return ((sticks % MAX_PLUS_ONE) + MAX) % MAX_PLUS_ONE;
    }

}

File: variants/kotlin-gradle/nim/aibest/src/main/kotlin/net/pijcke/nim/aibest/BestAi.kt
package net.pijcke.nim.aibest

import net.pijcke.nim.airandom.RandomAi
import net.pijcke.nim.rules.Nim
import net.pijcke.nim.rules.Player

class BestAi : Player {

    companion object {
        private const val MAX = Nim.MAX_STICKS_PER_TURN
        private const val MAX_PLUS_ONE = MAX + 1
    }

    private val fallbackStrategy = RandomAi()

    override fun name() = "Best AI"

    override fun play(sticks: Int) : Int {
        if (sticks == 1) return 1
        if (sticks % MAX_PLUS_ONE == 1) return fallbackStrategy.play(sticks)
        return ((sticks % MAX_PLUS_ONE) + MAX) % MAX_PLUS_ONE
    }

}
